### PR TITLE
Ticket #2228: Assign investigator to self

### LINKED ISSUE #2228: Analyst: "Assign to me" a request
### Issue description

PROBLEM: When an analyst chooses to assign a request to someone, they may type in some characters to search for themselves, or scroll thru the list of analysts to be able to find and select themselves.

SOLUTION: Since the most common action is to assign a request to themselves, a simple "assign to me" button will provide this action with a single click.

VALUE: This will streamline the analysts flow, and save time and fatigue.

Image

### Acceptance criteria

- [ ] An analysts may choose to assign an open request to themselves, by using a link that says "Assign to me," located to the right of the field and icons.
- [ ] The Option to "Assign to me" is unavailable if the issue is already assigned to someone (to avoid accidents)
- [ ] Design team has been consulted to resolve the design artifacts.

### Additional context

Originated from a comment by Cameron about usability in assigning the analyst. Discussed in Product/Design sync 5-28-24.

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #2228 

## Changes

- Adds a "assign to me" button under the investigator field in django admin (for domain request)
- Adds a small javascript hook to enable this behaviour

## Context for reviewers
The investigator field is a little different than our normal selectors: it uses the select2 library.
To trigger a change event on the field, we need to use the relevant jquery selector rather than .value

## Setup
1. Go to /admin
2. Click Domain Request
3. Select a record (with no investigator)
4. Click the assign to me button underneath the investigator field (see attached)

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots
image

### CODE REVIEW

**Review by kristinacyin — COMMENTED** [review]
Reviewing this right now! 

I noticed that the "Assign to me" button disappears as soon as an investigator is selected. I'm wondering if it might be a good idea to have the "Assign to me" button stay if the investigator selected isn't the user. It could be possible, for example, for a domain request to change hands from one investigator to another if it's a particularly tricky case. What do you think?

**Review by kristinacyin — DISMISSED** [review]
It seems at least that people are partial to my idea [on Slack](https://cisa-corp.slack.com/archives/C05BDEA3C11/p1718293395349429?thread_ts=1718217946.956789&cid=C05BDEA3C11), would you mind changing it so that the "Assign to me" button perpetuates if the investigator assigned is not the user? I'll also change the ticket AC, thank you!

**Review by kristinacyin — CHANGES_REQUESTED** [review]
Thank you for implementing the changes! I noticed that even when I have already assigned **myself** as the investigator, the "Assign to me" button still stays (screenshot below). I think that's the only state in which we can remove the "Assign to me" button, if possible. I don't have strong opinions about this so let me know if it's too much of a hassle

image

**Review by kristinacyin — APPROVED** [review]
Looks great! Great job!!! 🥇

File: src/registrar/assets/js/get-gov-admin.js [hunk 1]
```diff
@@ -137,6 +137,47 @@ function openInNewTab(el, removeAttribute = false){
...
+            // Create a DOM Option that matches the desired user. Then append it and select it.
+            let userOption = new Option(currentUserName, currentUserId, true, true);
+            selector.append(userOption).trigger("change");
+        }
+    });
+
+    // Listen to any change events, and hide the parent container if investigator has a value.
+    selector.on('change', function() {
+        // The parent container has display type flex.
+        assignSelfButton.parentElement.style.display = this.value === currentUserId ? "none" : "flex";
```
Comment (CocoByte) [inline src/registrar/assets/js/get-gov-admin.js:175]: Nice attention to style detail here.  Usually, I would view this as a symptom of design architecture needing an update, but given what we are working with I think this is perfectly fine.

**Review by CocoByte — APPROVED** [review]
LGTM

**Pull request merged**